gufe/protocols/protocoldag.py

```diff
@@ -0,0 +1,82 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/gufe
+
+import abc
+from typing import Iterable, List, Dict, Set, Optional
+
+import networkx as nx
+
+from .protocolunit import ProtocolUnit
+from .results import ProtocolDAGResult
+
+
+class ProtocolDAG:
+    """An executable directed, acyclic graph (DAG) composed of `ProtocolUnit`s
+    with dependencies specified.
+
+    This is the unit of execution passed to an alchemical `Scheduler`.
+    A `ProtocolDAG` yields a `ProtocolResult`, which can be placed in a `ResultStore`.
+
+    """
+
+    def __init__(
+        self,
+        graph: nx.DiGraph,
+        name: Optional[str] = None,
+    ):
+        """Create a new `ProtocolDAG`.
+
+        Parameters
+        ----------
+        graph : nx.DiGraph
+            The `ProtocolUnit`s, with dependencies set, as a networkx `DiGraph`.
+        name : str
+            Unique identifier for this `ProtocolDAG`.
+
+        """
+        self._graph = graph
+        self._name = name
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def graph(self):
+        return self._graph
+
+    def execute(self) -> ProtocolDAGResult:
+        """Execute the full DAG in-serial, in process."""
+
+        # operate on a copy, since we'll add ProtocolUnitResults as node attributes
+        graph = self._graph.copy(as_view=False)
+
+        completed: Set[ProtocolUnit] = set()
+
+        while len(completed) != len(self._graph.nodes):
+            for pu in graph.nodes:
```

Comment:
Pedantic but for the sake of sanity can we make `pu` a bit more explicitly named here? Someone is going to eventually mess up when looking at `pu` and `pur`

Reply:
reworked this loop to be much simpler